image-builder/image_builder.sh

```diff
@@ -76,55 +76,83 @@ OK "init is installed"
 OK "Agent installed"
 [ "$(id -u)" -eq 0 ] || die "$0: must be run as root"
 
+ROOTFS_SIZE=$(du -B 1MB -s "${ROOTFS}" | awk '{print $1}')
 BLOCK_SIZE=${BLOCK_SIZE:-4096}
-IMG_SIZE=${IMG_SIZE:-80}
-
-info "Creating raw disk with size ${IMG_SIZE}M"
-qemu-img create -q -f raw "${IMAGE}" "${IMG_SIZE}M"
-OK "Image file created"
-
-# Kata runtime expect an image with just one partition
-# The partition is the rootfs content
-
-info "Creating partitions"
-parted "${IMAGE}" --script "mklabel gpt" \
-"mkpart ext4 1M -1M"
-OK "Partitions created"
 
-# Get the loop device bound to the image file (requires /dev mounted in the
-# image build system and root privileges)
-DEVICE=$(losetup -P -f --show "${IMAGE}")
+# Extra buffer space for the raw disk in percentage.
+IMG_BUFFER=30
 
-#Refresh partition table
-partprobe "${DEVICE}"
+# Calculate image size based on the rootfs
+calculate_img_size()
+{
+    IMG_SIZE=${IMG_SIZE:-$(((($ROOTFS_SIZE*$IMG_BUFFER)/100)+$ROOTFS_SIZE))}
+}
 
-MOUNT_DIR=$(mktemp -d osbuilder-mount-dir.XXXX)
-info "Formating Image using ext4 format"
-mkfs.ext4 -q -F -b "${BLOCK_SIZE}" "${DEVICE}p1"
-OK "Image formated"
+# Cleanup
+cleanup()
+{
+    sync
+    umount -l ${MOUNT_DIR}
+    fsck -D -y "${DEVICE}p1"
+    losetup -d "${DEVICE}"
+}
 
-info "Mounting root paratition"
-mount "${DEVICE}p1" "${MOUNT_DIR}"
-OK "root paratition mounted"
+create_rootfs()
+{
+    calculate_img_size
+    info "Creating raw disk with size ${IMG_SIZE}M"
+    qemu-img create -q -f raw "${IMAGE}" "${IMG_SIZE}M"
+    OK "Image file created"
+
+    # Kata runtime expect an image with just one partition
+    # The partition is the rootfs content
+
+    info "Creating partitions"
+    parted "${IMAGE}" --script "mklabel gpt" \
+    "mkpart ext4 1M -1M"
```

Comment:
Might be worth making the FS type a variable which could default to `ext4`.

Reply:
Done. 

```
./image-builder/image_builder.sh -s 200 -f ext4 "${PWD}/fedora_rootfs"
Sending build context to Docker daemon  188.8MB
Step 1/2 : FROM fedora:latest
 ---> 422dc563ca32
Step 2/2 : RUN dnf install -y qemu-img parted gdisk e2fsprogs
 ---> Using cache
 ---> fa463f186699
Successfully built fa463f186699
Successfully tagged image-builder-osbuilder:latest
[OK] init is installed
[OK] Agent installed
INFO: Create root disk image. Attempt: 1
INFO: Creating raw disk with size 200M
INFO: Creating partitions
[OK] Image file created
[OK] Partitions created
INFO: Formating Image using ext4 format
INFO: Mounting root paratition
[OK] Image formated
[OK] root paratition mounted
INFO: Set filesystem reserved blocks percentage to 3%
tune2fs 1.43.5 (04-Aug-2017)
Setting reserved blocks percentage to 3% (1520 blocks)
INFO: Free space root partition  166 MB
INFO: rootfs size 133 MB
INFO: Copying content from rootfs to root partition
[OK] rootfs copied
fsck from util-linux 2.30.2
e2fsck 1.43.5 (04-Aug-2017)
/dev/loop3p1: clean, 6586/50688 files, 38172/50688 blocks
INFO: Image created. Size: 200MB.
```

Reply:
Defaults to `ext4`,

```
./image-builder/image_builder.sh -s 200  "${PWD}/fedora_rootfs"
Sending build context to Docker daemon  188.8MB
Step 1/2 : FROM fedora:latest
 ---> 422dc563ca32
Step 2/2 : RUN dnf install -y qemu-img parted gdisk e2fsprogs
 ---> Using cache
 ---> fa463f186699
Successfully built fa463f186699
Successfully tagged image-builder-osbuilder:latest
[OK] init is installed
[OK] Agent installed
INFO: Create root disk image. Attempt: 1
INFO: Creating raw disk with size 200M
[OK] Image file created
INFO: Creating partitions
[OK] Partitions created
INFO: Formating Image using ext4 format
[OK] Image formated
INFO: Mounting root paratition
[OK] root paratition mounted
INFO: Set filesystem reserved blocks percentage to 3%
tune2fs 1.43.5 (04-Aug-2017)
Setting reserved blocks percentage to 3% (1520 blocks)
INFO: Free space root partition  166 MB
INFO: rootfs size 133 MB
INFO: Copying content from rootfs to root partition
[OK] rootfs copied
fsck from util-linux 2.30.2
e2fsck 1.43.5 (04-Aug-2017)
/dev/loop3p1: clean, 6586/50688 files, 38172/50688 blocks
INFO: Image created. Size: 200MB.
```